Implement CanvasRenderingContext2D letterSpacing/wordSpacing

rwlbuis opened the pull request:
#### 00d61d715299f4a44a74385199c72f633cadee0d
<pre>
Implement CanvasRenderingContext2D letterSpacing/wordSpacing
<a href="https://bugs.webkit.org/show_bug.cgi?id=283408">https://bugs.webkit.org/show_bug.cgi?id=283408</a>

Reviewed by Said Abou-Hallawa.

Implement CanvasRenderingContext2D letterSpacing/wordSpacing [1, 2] by allowing to set these properties by string
on CanvasRenderingContext2D and as a computed Length on FontCascade. The CSS property parser is used to
convert the string value into the actual computed Length.

[1] <a href="https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-letterspacing">https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-letterspacing</a>
[2] <a href="https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-wordspacing">https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-wordspacing</a>

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/reset/2d.reset.state.letter_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/reset/2d.reset.state.word_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.absolute.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.font-relative.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.invalid.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.letterSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.letterSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.nonfinite.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.wordSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.wordSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.letter_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.letter_spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.word_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.word_spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.absolute.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.absolute.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.font-relative.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.font-relative.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.invalid.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.invalid.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.change.font.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.measure.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.nonfinite.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.nonfinite.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.change.font.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.measure.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/interfaces/html.idl:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/idlharness.worker-expected.txt:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::setFontWithoutUpdatingStyle):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::State::State):
(WebCore::CanvasRenderingContext2DBase::setLetterSpacing):
(WebCore::CanvasRenderingContext2DBase::setWordSpacing):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::letterSpacing const):
(WebCore::CanvasRenderingContext2DBase::wordSpacing const):
* Source/WebCore/html/canvas/CanvasTextDrawingStyles.idl:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
(WebCore::OffscreenCanvasRenderingContext2D::setFont):

Canonical link: <a href="https://commits.webkit.org/287661@main">https://commits.webkit.org/287661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7cdc199286416a291f35a29fd88e02e2f2c114a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80331 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59337 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84852 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7638 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62798 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20599 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83400 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52866 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73146 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43101 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50186 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27301 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29772 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71324 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27815 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86286 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7556 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5348 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71082 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7731 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68982 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70322 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17531 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14305 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13253 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7518 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13038 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7357 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10877 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9163 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
```diff
@@ -214,6 +214,11 @@ void CanvasRenderingContext2D::setFontWithoutUpdatingStyle(const String& newFont
     ASSERT(state().font.realized());
     ASSERT(state().font.isPopulated());
+
+    String letterSpacing;
+    setLetterSpacing(std::exchange(modifiableState().letterSpacing, letterSpacing));
+    String wordSpacing;
+    setWordSpacing(std::exchange(modifiableState().wordSpacing, wordSpacing));
```
Please add a comment here saying that we need to recompute the word and the letter spacing for the new font.

rwlbuis replied:
Done.

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -3060,4 +3064,38 @@ std::optional<CanvasRenderingContext2DBase::RenderingMode> CanvasRenderingContex
+    CSSTokenizer tokenizer(letterSpacing);
+    auto tokenRange = tokenizer.tokenRange();
+    tokenRange.consumeWhitespace();
+    RefPtr<CSSPrimitiveValue> parsedValue = CSSPropertyParserHelpers::consumeLength(tokenRange, HTMLStandardMode);
```
I think this can be just 
```
RefPtr lengthValue = ...
```
Since it is the return of `consumeLength()`

rwlbuis replied:
Sure, fixed.

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -3060,4 +3064,38 @@ std::optional<CanvasRenderingContext2DBase::RenderingMode> CanvasRenderingContex
+    auto tokenRange = tokenizer.tokenRange();
+    tokenRange.consumeWhitespace();
+    RefPtr<CSSPrimitiveValue> parsedValue = CSSPropertyParserHelpers::consumeLength(tokenRange, HTMLStandardMode);
+    if (!parsedValue || !parsedValue->isLength())
```
I do not think we need to check `!parsedValue->isLength()`.

rwlbuis replied:
Removed.

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
```diff
@@ -264,6 +270,14 @@ class CanvasRenderingContext2DBase : public CanvasRenderingContext, public Canva
         bool isPopulated() const { return m_font.fonts(); }
 #endif
 
+        FontCascade fontCascade() const { return m_font; }
```
I think this should return `const FontCascade&` instead of copying the object.

rwlbuis replied:
Done.

shallawa commented on Source/WebCore/html/canvas/CanvasTextDrawingStyles.idl:
```diff
@@ -30,4 +30,6 @@ interface mixin CanvasTextDrawingStyles {
     attribute CanvasTextBaseline textBaseline; // (default: "alphabetic")
     attribute CanvasDirection direction; // (default: "inherit")
+    attribute DOMString letterSpacing;
+    attribute DOMString wordSpacing;
```
Should we state the default values here like the rest of the members above:
```
    attribute DOMString letterSpacing;  // (default: "0px")
    attribute DOMString wordSpacing;  // (default: "0px")
```

rwlbuis replied:
Yes, should have noticed the other comments, fixed.

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -3060,4 +3064,38 @@ std::optional<CanvasRenderingContext2DBase::RenderingMode> CanvasRenderingContex
+    if (!parsedValue || !parsedValue->isLength())
+        return;
+    modifiableState().letterSpacing = parsedValue->cssText();
+    auto fontCascade = fontProxy()->fontCascade();
```
I think this should be referencing the fontCascade()
```
auto& fontCascade = ...
```

rwlbuis replied:
Done.

rwlbuis committed:
Implement CanvasRenderingContext2D letterSpacing/wordSpacing
https://bugs.webkit.org/show_bug.cgi?id=283408

Reviewed by Said Abou-Hallawa.

Implement CanvasRenderingContext2D letterSpacing/wordSpacing [1, 2] by allowing to set these properties by string
on CanvasRenderingContext2D and as a computed Length on FontCascade. The CSS property parser is used to
convert the string value into the actual computed Length.

[1] https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-letterspacing
[2] https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-wordspacing

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/reset/2d.reset.state.letter_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/reset/2d.reset.state.word_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.absolute.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.font-relative.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.invalid.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.letterSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.letterSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.nonfinite.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.wordSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.drawing.style.wordSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.letter_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.letter_spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.word_spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/reset/2d.reset.state.word_spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.absolute.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.absolute.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.font-relative.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.font-relative.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.invalid.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.invalid.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.change.font.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.letterSpacing.measure.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.nonfinite.spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.nonfinite.spacing.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.change.font-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.change.font.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.measure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.drawing.style.wordSpacing.measure.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/interfaces/html.idl:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/idlharness.worker-expected.txt:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::setFontWithoutUpdatingStyle):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::State::State):
(WebCore::CanvasRenderingContext2DBase::setLetterSpacing):
(WebCore::CanvasRenderingContext2DBase::setWordSpacing):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::letterSpacing const):
(WebCore::CanvasRenderingContext2DBase::wordSpacing const):
* Source/WebCore/html/canvas/CanvasTextDrawingStyles.idl:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
(WebCore::OffscreenCanvasRenderingContext2D::setFont):

Canonical link: https://commits.webkit.org/287661@main
